src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,192 @@
+    {
+        const ValueType reciWidth = float_X(1.0) / CELL_WIDTH;
+        const ValueType reciHeight = float_X(1.0) / CELL_HEIGHT;
+        return ((*field(1,0)).x() - (*field).x()) * reciWidth +
+               ((*field(0,1)).y() - (*field).y()) * reciHeight;
+    }
+};
+
+// functor for all species to calculate density
+template<typename T_SpeciesName, typename T_Area>
+struct ComputeChargeDensity
+{
+    typedef typename T_SpeciesName::type SpeciesName;
+    static const uint32_t area = T_Area::value;
+
+    HINLINE void operator()( FieldTmp* fieldTmp,
+                             const uint32_t currentStep) const
+    {
+        DataConnector &dc = Environment<>::get().DataConnector();
+        
+        /* load species without copying the particle data to the host */
+        SpeciesName* speciesTmp = &(dc.getData<SpeciesName >(SpeciesName::FrameType::getName(), true));
+
+        /* run algorithm */
+        typedef typename CreateDensityOperation<SpeciesName>::type::Solver ChargeDensitySolver;
+        fieldTmp->computeValue < area, ChargeDensitySolver > (*speciesTmp, currentStep);
+        dc.releaseData(SpeciesName::FrameType::getName());
+    }
+};
+
+} // namespace detail
+
+void ChargeConservation::notify(uint32_t currentStep)
+{
+    typedef SuperCellSize BlockDim;
+
+    DataConnector &dc = Environment<>::get().DataConnector();
+
+    /* load FieldTmp without copy data to host */
+    FieldTmp* fieldTmp = &(dc.getData<FieldTmp > (FieldTmp::getName(), true));
+    /* reset density values to zero */
+    fieldTmp->getGridBuffer().getDeviceBuffer().setValue(FieldTmp::ValueType(0.0));
+
+    /* calculate and add the charge density values from all species in FieldTmp */
+    ForEach<VectorAllSpecies, detail::ComputeChargeDensity<bmpl::_1,bmpl::int_<CORE + BORDER> >, MakeIdentifier<bmpl::_1> > computeChargeDensity;
+    computeChargeDensity(forward(fieldTmp), currentStep);
+
+    /* add results of all species that are still in GUARD to next GPUs BORDER */
+    EventTask fieldTmpEvent = fieldTmp->asyncCommunication(__getTransactionEvent());
+    __setTransactionEvent(fieldTmpEvent);
+    
+    /* cast libPMacc Buffer to cuSTL Buffer */
+    BOOST_AUTO(fieldTmp_coreBorder,
+                 fieldTmp->getGridBuffer().
+                 getDeviceBuffer().cartBuffer().
+                 view(BlockDim::toRT(), -BlockDim::toRT()));
+    
+    /* cast libPMacc Buffer to cuSTL Buffer */
+    BOOST_AUTO(fieldE_coreBorder,
+                 dc.getData<FieldE > (FieldE::getName(), true).getGridBuffer().
+                 getDeviceBuffer().cartBuffer().
+                 view(BlockDim::toRT(), -BlockDim::toRT()));
+
+    /* run calculation: fieldTmp = | div E - rho / eps_0 | */
+    using namespace lambda;
+    using namespace PMacc::math::math_functor;
+    typedef detail::Div<simDim, typename FieldTmp::ValueType> myDiv;
+    algorithm::kernel::Foreach<BlockDim>()
+        (fieldTmp_coreBorder.zone(), fieldTmp_coreBorder.origin(), 
```

Comment:
In this case `fieldE_coreBorder` shape is N_x/2, N_y/2, N_z/2. Nevertheless the allocated memory block underneath is larger, because of the guard-area. It is the same as in libpmacc